fix: allow unsupported Zod types in tool outputSchema

camachoh1 opened the pull request:
## Summary

Fixes #8040 - Tools can now use any valid Zod type (including `ZodTuple`, `ZodIntersection`, etc.) in their `outputSchema` without triggering model compatibility errors.

## Problem

When creating a tool with `ZodTuple` in the `outputSchema`, the following error was thrown:
```
Error: openai/gpt-4.1-mini does not support zod type: ZodTuple
```

This happened because the tool builder was applying model-specific compatibility validation to `outputSchema`, even though **outputSchema is never sent to the LLM** - it's only used internally by Mastra for validating tool execution results.

## Solution

Modified `CoreToolBuilder.build()` in `packages/core/src/tools/tool-builder/builder.ts` to skip model compatibility layers for `outputSchema`:

**Before:**
```typescript
processedOutputSchema = applyCompatLayer({
  schema: this.getOutputSchema(),
  compatLayers: schemaCompatLayers,  // Applied model validation ❌
  mode: 'aiSdkSchema',
});
```

**After:**
```typescript
// Don't apply model compatibility layers to outputSchema since it's never sent to the LLM
processedOutputSchema = convertZodSchemaToAISDKSchema(this.getOutputSchema());
```

## Testing

Added comprehensive tests that verify:
1. ✅ `outputSchema` with `ZodTuple` no longer throws errors
2. ✅ `outputSchema` with `ZodIntersection` works correctly
3. ✅ `inputSchema` with `ZodTuple` still correctly throws errors (validation preserved)

All tests pass.

## Why This is Safe

- `outputSchema` is only used for internal validation of tool execution results
- The LLM only receives: `name`, `description`, and `parameters` (inputSchema)
- `inputSchema` still has full model compatibility validation
- This change aligns with the AI SDK's tool specification

Co-Authored-By: Richard LoRicco <rtlo9962@gmail.com>

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Tool builds no longer apply compatibility-layer transformations to tool output schemas, so complex output types (e.g., tuples, intersections) are preserved and no longer trigger compatibility-related failures.

* **Tests**
  * Added a test ensuring complex output schemas are retained and exposed correctly when a tool is built.

* **Chores**
  * Added a changeset documenting the output schema validation fix.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Skips model compatibility on tool `outputSchema` so unsupported Zod types (e.g., `ZodTuple`) are allowed; adds a test and a changeset.
> 
> - **Core**
>   - Tool builder (`packages/core/src/tools/tool-builder/builder.ts`): do not apply schema compatibility layers to `outputSchema` (never sent to LLM); input schema handling unchanged.
> - **Tests**
>   - Add test in `builder.test.ts` verifying `outputSchema` can include `ZodTuple` without errors and remains defined after build.
> - **Docs/Release**
>   - Add changeset (`.changeset/fix-output-schema-zod-tuple.md`) documenting the validation fix.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d12dff44b8c9a490b41def87bcac31814656d56c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

camachoh1 committed:
test: add failing test reproducing outputSchema tuple validation bug

DanielSLew commented on packages/core/src/tools/tool-builder/builder.ts:
```diff
@@ -391,11 +391,10 @@ export class CoreToolBuilder extends MastraBase {
+      // Don't apply model compatibility layers to outputSchema since it's never sent to the LLM
+      // The outputSchema is only used internally for validation, so it can contain any valid Zod type
+      // including ZodTuple, ZodIntersection, etc. that some models don't support in their input schemas
+      processedOutputSchema = convertZodSchemaToAISDKSchema(this.getOutputSchema());
```
```suggestion
      // Don't add any compat layers to outputSchema since it's never sent to the LLM
      processedOutputSchema = applyCompatLayer({
        schema: this.getOutputSchema(),
        compatLayers: [],
        mode: 'aiSdkSchema'
      });
```

camachoh1 committed:
fix: skip model compatibility validation for outputSchema

outputSchema is never sent to the LLM so it doesn't need model-specific
validation. This allows tools to use any valid Zod type (including
ZodTuple, ZodIntersection, etc.) in their output schemas.

Fixes #8040

Co-Authored-By: Richard LoRicco <rtlo9962@gmail.com>

DanielSLew committed:
Merge branch 'main' into fix/output-schema-tuple-validation

camachoh1 committed:
refactor: use applyCompatLayer with empty layers for outputSchema

Apply reviewer feedback to use applyCompatLayer with an empty
compatLayers array instead of calling convertZodSchemaToAISDKSchema
directly. This maintains consistency with the existing architecture
while still bypassing model compatibility checks for outputSchema.

DanielSLew committed:
Merge branch 'main' into fix/output-schema-tuple-validation

camachoh1 committed:
fix: resolve linting error for unused context parameter

camachoh1 committed:
fix: add type assertion to resolve TypeScript type instantiation error

DanielSLew committed:
Merge branch 'main' into fix/output-schema-tuple-validation

DanielSLew commented on packages/core/src/tools/tool-builder/builder.test.ts:
```diff
@@ -949,3 +949,49 @@ describe('Tool Input Validation', () => {
+        conditions: z.string(),
+        test: z.tuple([z.string(), z.string()]),
+      }),
+      execute: async ({ context: _ }) => ({
```
you can just remove variables if they're not being used
```suggestion
      execute: async () => ({
```

camachoh1 committed:
Merge remote-tracking branch 'upstream/main' into fix/output-schema-tuple-validation

camachoh1 committed:
refactor: address reviewer feedback on output schema test

- Remove unused execute parameter (context)
- Remove unnecessary 'as any' cast from originalTool
- Rename runtimeContext to requestContext for consistency

camachoh1 committed:
Merge branch 'main' into fix/output-schema-tuple-validation

DanielSLew committed:
Merge branch 'main' into fix/output-schema-tuple-validation

camachoh1 committed:
Add changeset for output schema fix

DanielSLew committed:
Merge branch 'main' into fix/output-schema-tuple-validation

DanielSLew committed:
Merge branch 'main' into fix/output-schema-tuple-validation
